src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +438,154 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
+                                           &d_registrar,
+                                           &d_context,
+                                           true,
+                                           "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  if (solveByCryptominisat(&d_cnfStream, assertionsToPreprocess, substs_index)
+      == SAT_VALUE_FALSE)
+  {
+    // If in conflict, just return false
+    return std::make_tuple(false, learned_literals);
+  }
+
+  std::vector<SatLiteral> topLevelUnits = d_satSolver->getTopLevelUnits();
+  for (size_t i = 0; i < topLevelUnits.size(); i++)
+  {
+    SatLiteral lit = topLevelUnits[i];
+    if (d_cnfStream.getNodeCache().find(lit)
+        != d_cnfStream.getNodeCache().end())
+    {
+      // if the literal is in the getNodeCache
+      Node learnedLiteral = Rewriter::rewrite(d_cnfStream.getNode(lit));
+      if (learnedLiteral.isConst())
+      {
+        if (learnedLiteral.getConst<bool>())
+        {
+          // If the learned literal simplifies to true, it's redundant
+          continue;
+        }
+        else
+        {
+          // If the learned literal simplifies to false, we're in conflict
+          return std::make_tuple(false, learned_literals);
+        }
+      }
+      learned_literals.push_back(learnedLiteral);
+    }
+  }
+  return std::make_tuple(true, learned_literals);
+}
+
+bool NonClausalSimp::solveByCircuitPropagator(
+    theory::booleans::CircuitPropagator* propagator,
+    AssertionPipeline* assertionsToPreprocess,
+    unsigned substs_index)
+{
+  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    Trace("non-clausal-simplify") << "Assertion #" << i << " : "
+                                  << (*assertionsToPreprocess)[i] << std::endl;
+  }
+
+  if (propagator->getNeedsFinish())
+  {
+    propagator->finish();
+    propagator->setNeedsFinish(false);
+  }
+  propagator->initialize();
+
+  // Assert all the assertions to the propagator
+  Trace("non-clausal-simplify") << "asserting to propagator" << std::endl;
+
+  { // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_addingAssertions);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("non-clausal-simplify")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      Debug("cores") << "propagator->assertTrue: " << (*assertionsToPreprocess)[i]
+                     << std::endl;
+      propagator->assertTrue((*assertionsToPreprocess)[i]);
+    }
+  } // end timing
+
+  Trace("non-clausal-simplify") << "propagating" << std::endl;
+  return propagator->propagate();
+}
+
+std::tuple<bool, std::vector<Node>> NonClausalSimp::preprocessByCircuitPropagator(
```

Comment:
It looks like these `preprocess*` methods are not doing that much (especially if we simplify the rest according to my other comments), so maybe they are not necessary and can just be merged with the `solve*` methods.

Reply:
These methods are responsible for getting the learned clauses. Do you mean putting the solveBy* method in the preprocessing method?

Reply:
Yeah, exactly. If they are not really doing anything besides calling the `solveBy*` methods, they can be inlined.

Reply:
I just merged the preprocess and solve methods to one because neither of them is that long.